### ISSUE TITLE
Right to Left implemention 

### ISSUE THREAD

**Issue by aimanmutasem** [bentrevett/pytorch-seq2seq#98]
Hi Ben, :)

I'm training to implement your code from right to left. 

I have tried to reverse the input sequence from (L to R) to (R to L), but I think this is not a good idea. According to your experience, what is the optimal way to implement notebook 6 (Transformer) from right to left? 

Sincerely,
Aiman

**Comment by bentrevett** [comment]
Reversing the input sequence was only done in the first notebook as it implements a paper in which they reverse the input sequence as they believe it helps their model. From some quick experiments I have found it does not really provide any performance increase in the other notebooks so should probably be left as left-to-right.

**Comment by aimanmutasem** [comment]
You are right, already I have got lower results. My intention is to apply the transformer with two models (RTL and LTR) simultaneously. 

Kindly, take a look at the attached image and give me your opinion about how to implement something like this. 

![Image_20200416215705](https://user-images.githubusercontent.com/7637767/79465475-f75a5f00-802d-11ea-9073-028c135c578a.jpg)

**Comment by aimanmutasem** [comment]
Hi Ben, 

Really, I'm so sorry to post again and again, and I know this is not an issue. You're an unselfish developer and helped me a lot, I appreciate. 

 I have applied two ideas:

1. reversing the input text from left to Right, before feeding the data to the encoder and decoder. (but the training process is still left to right)

2. reversing the (embedding vector) in the encoder and decoder. (the result is so boor, and the BLUE score is decreased by 15.6 points)

I hope to support me to find the optimal way to train model Right to Left.

Note: I have two versions of my model (RTL model) and (LTR model).

**Comment by bentrevett** [comment]
Don't worry about posting questions - that's what I want to use the issue section for.

Are you working with a language that is usually right-to-left?

**Comment by aimanmutasem** [comment]
Thank you so much for your time :) 

I'm Arabic native speaker. The current project is to apply the basic transformer in two directions RTL and LTR (two models). Then boosting each other.  It will work well for Arabic and English.

**Comment by aimanmutasem** [comment]
Hi Ben, this Aiman Again :)

I get feedback from a friend like : 

"You need not only reverse the embedding vector of the decoder, but also the predicted target tokens, so as that the t token can predict the t-1 token in the target sentence"

Accordingly, I have applied as bellow:

For reversing embedding decoder : 

`        pos = torch.arange(0, trg_len).unsqueeze(0).repeat(batch_size, 1).to(self.device)
                            
        #pos = [batch size, trg len]
            
        trg = self.dropout((self.tok_embedding(trg) * self.scale) + self.pos_embedding(pos))
        
        idx = [i for i in range(trg.size(0)-1, -1, -1)]
        idx = torch.LongTensor(idx).to(self.device)
        trg = torch.index_select(trg,0,idx)

        #trg = [batch size, trg len, hid dim]`

For predicted target tokens in the training section I have applied as : 

`output = output.contiguous().view(-1, output_dim)
        trg = trg[:,1:].contiguous().view(-1)
        
        #idx = [i for i in range(trg.size(0)-1, -1, -1)]
        #idx = torch.LongTensor(idx).to(device)
        #trg = torch.index_select(trg,0,idx)        
`
but the performance is still poor do you think I'm applied it correct ? 
Please, check the below link about how to reverse tensoe: 
https://discuss.pytorch.org/t/how-to-reverse-a-torch-tensor/382/5

Sincerely,
